cmd/kubeadm/app/util/staticpod/utils.go

```diff
@@ -406,3 +410,29 @@ func DeepHashObject(hasher hash.Hash, objectToWrite interface{}) {
 	hasher.Reset()
 	fmt.Fprintf(hasher, "%v", dump.ForHash(objectToWrite))
 }
+
+// DiffReporter is a simple custom reporter that only records differences
+// detected during comparison.
+type DiffReporter struct {
+	path  cmp.Path
+	diffs []string
+}
+
+func (r *DiffReporter) PushStep(ps cmp.PathStep) {
+	r.path = append(r.path, ps)
+}
+
+func (r *DiffReporter) Report(rs cmp.Result) {
+	if !rs.Equal() {
+		vx, vy := r.path.Last().Values()
+		r.diffs = append(r.diffs, fmt.Sprintf("%#v:\n\t-\t\t\t\t\t%+v\n\t+\t\t\t\t\t%+v\n", r.path, vx, vy))
```

Comment:
i must say i don't like this or paco's string replace solition. isn't there something stock we can use?

how about marshling to yaml / json and checking a string diff? is that deterministic?

Reply:
Test convert Pod to Yaml first:
```go
	m1, _ := kubeadmutil.MarshalToYaml(pod1, v1.SchemeGroupVersion)
	m2, _ := kubeadmutil.MarshalToYaml(pod2, v1.SchemeGroupVersion)
	return false, cmp.Diff(string(m2), string(m1)), nil
```
But it seems that the output is not very readable :(
```
            	actual diff:   strings.Join({
              	... // 97 identical bytes
              	"er: control-plane\n  name: etcd\n  namespace: kube-system\nspec:\n  ",
              	"containers:\n  - image: gcr.io/google_containers/etcd-amd64:3.1.1",
            - 	"2",
            + 	"1",
              	"\n    name: \"\"\n    resources: {}\nstatus: {}\n",
              }, "")

...
            	actual diff:   (
              	"""
              	... // 12 identical lines
              	    name: ""
              	    resources: {}
            - 	  restartPolicy: Always
              	status: {}
              	"""
              )
```
Is there a good way to format the output?
Or maybe we just do not need to check the specific content of the diff, but check whether `diff != ""`?

Reply:
how about using `github.com/pmezard/go-difflib/difflib` ?
see app/cmd/upgrade/diff.go

Reply:
but also we use cmp.Diff a lot:

```
~/go/src/k8s.io/kubernetes/cmd/kubeadm$ grep cmp.Diff * -rnI
app/phases/etcd/local_test.go:213:					cmp.Diff(string(manifestBytes), test.expectedManifest),
app/util/staticpod/utils.go:377:	return false, cmp.Diff(pod2, pod1), nil
app/cmd/init_test.go:136:				if diff := cmp.Diff(validData, data, cmp.AllowUnexported(initData{}), cmpopts.IgnoreFields(initData{}, "client", "cfg.ClusterConfiguration", "cfg.NodeRegistration.Taints")); diff != "" {
app/cmd/reset_test.go:236:				if diff := cmp.Diff(tc.data, data, cmp.AllowUnexported(resetData{}), cmpopts.IgnoreFields(resetData{}, "client", "cfg")); diff != "" {
app/cmd/join_test.go:243:				if diff := cmp.Diff(validData, data, cmp.AllowUnexported(joinData{}), cmpopts.IgnoreFields(joinData{}, "client", "initCfg", "cfg.ControlPlane.LocalAPIEndpoint")); diff != "" {
app/preflight/checks_test.go:1057:			if diff := cmp.Diff(checkList, rt.expStr); diff != "" {
app/preflight/checks_test.go:1124:			if diff := cmp.Diff(checkList, rt.expStr); diff != "" {

```

i am confused how other tests are not failing.

Reply:
/cc @pacoxu

Reply:
here is my proposal:

```patch
diff --git a/cmd/kubeadm/app/util/staticpod/utils.go b/cmd/kubeadm/app/util/staticpod/utils.go
index 9e9b75738ef..3e5f769a8a7 100644
--- a/cmd/kubeadm/app/util/staticpod/utils.go
+++ b/cmd/kubeadm/app/util/staticpod/utils.go
@@ -29,8 +29,8 @@ import (
 	"strings"
 	"sync"
 
-	"github.com/google/go-cmp/cmp"
 	"github.com/pkg/errors"
+	"github.com/pmezard/go-difflib/difflib"
 
 	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/resource"
@@ -374,7 +374,24 @@ func ManifestFilesAreEqual(path1, path2 string) (bool, string, error) {
 	if bytes.Equal(hash1, hash2) {
 		return true, "", nil
 	}
-	return false, cmp.Diff(pod2, pod1), nil
+
+	pod1Yaml, err := kubeadmutil.MarshalToYaml(pod1, v1.SchemeGroupVersion)
+	if err != nil {
+		return false, "", err
+	}
+	pod2Yaml, err := kubeadmutil.MarshalToYaml(pod2, v1.SchemeGroupVersion)
+	if err != nil {
+		return false, "", err
+	}
+
+	var buf bytes.Buffer
+	diff := difflib.UnifiedDiff{
+		A: difflib.SplitLines(string(pod1Yaml)),
+		B: difflib.SplitLines(string(pod2Yaml)),
+	}
+	difflib.WriteUnifiedDiff(&buf, diff)
+
+	return false, buf.String(), nil
 }
 
 // getProbeAddress returns a valid probe address.
diff --git a/cmd/kubeadm/app/util/staticpod/utils_test.go b/cmd/kubeadm/app/util/staticpod/utils_test.go
index 6fadfb3e1eb..9587b93e02c 100644
--- a/cmd/kubeadm/app/util/staticpod/utils_test.go
+++ b/cmd/kubeadm/app/util/staticpod/utils_test.go
@@ -766,18 +766,18 @@ func TestManifestFilesAreEqual(t *testing.T) {
 			podYamls:       []string{validPod, validPod2},
 			expectedResult: false,
 			expectErr:      false,
-			expectedDiff: string(`
-- 					"2",
-+ 					"1",`),
+			expectedDiff: string(`@@ -12 +12 @@
+-  - image: gcr.io/google_containers/etcd-amd64:3.1.11
++  - image: gcr.io/google_containers/etcd-amd64:3.1.12
+`),
 		},
 		{
-			description:    "manifests are not equal for adding new defaults",
+			description:    "manifests are not equal after adding new defaults",
 			podYamls:       []string{validPod, invalidWithDefaultFields},
 			expectedResult: false,
 			expectErr:      false,
-			expectedDiff: string(`
-- 		RestartPolicy:                 "Always",
-+ 		RestartPolicy:                 "",`),
+			expectedDiff: string(`@@ -14,0 +15 @@
++  restartPolicy: Always`),
 		},
 		{
 			description:    "first manifest doesn't exist",
@@ -830,7 +830,7 @@ func TestManifestFilesAreEqual(t *testing.T) {
 			}
 			if !strings.Contains(diff, rt.expectedDiff) {
 				t.Errorf(
-					"ManifestFilesAreEqual diff doesn't expected\n%s\n\texpected diff: %s\nactual diff: %s",
+					"ManifestFilesAreEqual diff doesn't expected\n%s\n\texpected diff:\n%s\nactual diff:\n%s",
 					rt.description,
 					rt.expectedDiff,
 					diff,
```

Reply:
Updated using the difflib in #120601.